### PR TITLE
Add MetricsSystem resetAll logic

### PR DESCRIPTION
**Pull request by LuQQiu**
Add ResetAllMetrics logic to the existing MetricsSystem.

### CODE REVIEW

**Comment by LuQQiu** [comment]
alluxio.master.table.transform.TransformManagerTest.jobHistory(alluxio.master.table.transform.TransformManagerTest)
  Run 1: TransformManagerTest.jobHistory:223->heartbeat:328 » Runtime Timer for thread ...
  Run 2: TransformManagerTest.jobHistory:223->heartbeat:328 » Runtime Timer for thread ...

**Comment by LuQQiu** [comment]
jenkins, test this please

**Comment by LuQQiu** [comment]
@calvinjia PTAL, I mainly add the resetAllMetrics logic in the MetricsSystem. Thanks

File: core/common/src/main/java/alluxio/metrics/MetricsSystem.java [hunk 1]
```diff
@@ -531,15 +532,59 @@ public static void resetCountersAndGauges() {
...
+    // Counters can be reset to zero values.
+    for (Counter counter : METRIC_REGISTRY.getCounters().values()) {
+      counter.dec(counter.getCount());
+    }
+
+    // No reset logic exist in Meter, a remove and add combination is needed
+    for (String meterName : METRIC_REGISTRY.getMeters().keySet()) {
+      METRIC_REGISTRY.remove(meterName);
+      METRIC_REGISTRY.meter(meterName);
+    }
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/metrics/MetricsSystem.java]: Will this cause a problem if other threads are concurrently updating the metric?

File: core/common/src/main/java/alluxio/metrics/MetricsSystem.java [hunk 2]
```diff
@@ -531,15 +532,59 @@ public static void resetCountersAndGauges() {
...
+  /**
+   * Resets all the metrics in the MetricsSystem.
+   *
+   * This method is not thread-safe and should be used sparingly.
+   */
+  public static void resetAllMetrics() {
+    if (sIsResetting) {
+      return;
+    }
+    sIsResetting = true;
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply core/common/src/main/java/alluxio/metrics/MetricsSystem.java]: The `sIsResetting` only guarantees that there is only one thread resetting the MetricsSsystem.

[the remark this answers is not in the document]
Comment (calvinjia) [reply core/common/src/main/java/alluxio/metrics/MetricsSystem.java]: This is not the case currently right? Two threads can evaluate line 541 as false and continue concurrently?

File: core/common/src/main/java/alluxio/metrics/MetricsSystem.java (hunk 1 above)
Comment (LuQQiu) [reply core/common/src/main/java/alluxio/metrics/MetricsSystem.java]: I previously consider adding locking mechanism for accessing `METRIC_REGISTRY` or use `waitForResetFinish` in MetricsSystem.counter/meter/timer/allMetrics() function. All those ways to guarantee correctness will add some kind of latency when accessing the Metrics. Many of our metrics are added or calculated in the critical path and will affect the performance of Alluxio RPC/read/write.  The metrics resetAll logics is not important enough for us to impact Alluxio overall performance.

For the safeness part, 
1) counters are mostly incremental. `counter.dec(counter.getCount());` is safe.
For the metrics `LAST_BACKUP_ENTRIES_COUNT` and `LAST_BACKUP_RESTORE_COUNT` which use `counter.dec(counter.getCount())`, I will change them to be Metric of type `Gauge`. They are not actual counter.

2) For meter and timer, we have remove and add operation combination.
```
METRIC_REGISTRY.remove(meterName);
METRIC_REGISTRY.meter/timer(meterName);
```
For remove operation, it may be concurrently with getOrAdd() operation. If the get() is faster than remove(),  we will update the old meter/timer instead of the new one. Some metrics data may be lost, but I don't think it's a big problem.  

MetricsRegistry.meter/timer() operation are thread-safe. They internally implement getOrAdd() function. It will only register a metric for once and use concurrentMap to guarantee that.

To summarize, add locking or wait for reset finish logics will be more correct, but the metrics system may add latency to Alluxio operations. Considering that, I will prefer to tolerate the small bias of the metrics data.

File: core/common/src/main/java/alluxio/metrics/MetricsSystem.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (LuQQiu) [reply core/common/src/main/java/alluxio/metrics/MetricsSystem.java]: Changed to use synchronized block, thanks

**Comment by calvinjia** [comment]
alluxio-bot, merge this please.
